### ISSUE TITLE
Make -obfl-keep-with-next-sheets work across sequences

### ISSUE THREAD

**Issue by bertfrees** [nlbdev/pipeline#91]
See [nlb-default.scss#L226](https://github.com/nlbdev/pipeline/blob/4e5a8ed/modules/nlb/book-to-pef/src/main/resources/css/nlb-default.scss#L226).

**Comment by bertfrees** [comment]
I've checked that the intermediate OBFL file does have the `keep-with-next-sheets` attributes, so the issue can't be in the css-to-obfl part. I also unit test `keep-with-next-sheets` so that proves it works at least in that case. I'd like to see an example of a case where it doesn't have the desired effect.

**Comment by bertfrees** [comment]
@josteinaj I think this can be closed? 17362e6614954908691c4e878b2aa4681871e9d3

**Comment by josteinaj** [comment]
Ok, sure. We ended up not using it I think so I don't know if we can test it.

**Comment by bertfrees** [comment]
This fix got dropped when updating to the latest Dotify because the patch was not accepted and we ended up not needing it (see https://github.com/nlbdev/pipeline/issues/88#issuecomment-342846757). If needed I can rebase https://github.com/nlbdev/pipeline/commit/17362e6614954908691c4e878b2aa4681871e9d3.

**Comment by josteinaj** [comment]
@bertfrees I'm trying to use `-obfl-keep-with-next-sheets` to handle orphaned headlines. I haven't got it working yet. I suppose the issue is that `-obfl-keep-with-next-sheets` doesn't work across OBFL sequences?

I've added a `keep-with-next-section` class using XSLT and will add `-obfl-keep-with-next-sheets: 1;` for elements inside a div/section with that class in the SCSS. The HTML should now look like this (i.e. just with an added `keep-with-next-section` class):

```html
<section id="level1_9" epub:type="part">
    <div class="leaf-section keep-with-next-section">
        <div epub:type="pagebreak" id="page-31" class="page-normal">31</div>
        <h1 id="h1_6">DEL 1</h1>
        <div epub:type="pagebreak" id="page-32" class="page-normal">32</div>
    </div>
    <section id="level2_6" class="chapter">...</section>
</section>
```

I'm trying this SCSS:
```scss
*.keep-with-next-section {
    > *:first-child, > h1, > h2, > h3, > h4, > h5, > h6 {
        -obfl-keep-with-next-sheets: 1;
    }
}
```

In the OBFL, the `keep-with-next-sheets` attribute is applied to the headline:
```xml
                    <block align="center"
                           border-bottom-align="outer"
                           border-bottom-style="solid"
                           border-left-align="outer"
                           border-left-style="solid"
                           border-right-align="outer"
                           border-right-style="solid"
                           border-top-align="outer"
                           border-top-style="solid"
                           id="d29987e564"
                           keep-with-next-sheets="1"
                           margin-bottom="1"
                           margin-left="1"
                           margin-right="5"
                           margin-top="2">
                        <block>
                            <marker class="print-page/prev" value="30"/>
                            <marker class="print-page" value="​"/>
                            <marker class="print-page/prev" value="​"/>
                            <marker class="print-page" value="⠿31"/>
                            <marker class="footer/prev" value="Hvem har ansvaret?"/>
                            <marker class="footer" value="DEL 1"/>
                        </block>
                        <marker class="indicator/pagenum" value="x"/>
                        ⠠⠠⠙⠑⠇ ⠼⠁
                        <marker class="print-page/prev" value="⠿31"/>
                        <marker class="print-page" value="⠿32"/>
                        <marker class="indicator/pagenum" value="x"/>
                    </block>
                </block>
            </block>
        </block>
    </block>
</sequence>
<sequence master="master_4" page-number-counter="page">
```

The headline is still orphaned even though it has `keep-with-next-sheets="1"` in the OBFL.

@bertfrees Was there a reason why the patch was not accepted into Dotify? This sounds like a bug in Dotify. Could you make `keep-with-next-sheets` work across sequences again?

**Comment by bertfrees** [comment]
Yes I think I can make it work across sequences again, but I don't know if it's such a good idea to do it if Joel is not convinced keep-with-next-sheets should behave like that. We have discussed it in the context of https://github.com/brailleapps/dotify.formatter.impl/pull/28, but I don't remember exactly what Joel's arguments were. Shall I bring it up with him again?

**Comment by josteinaj** [comment]
Yes, I think it's best if you bring it up with him again.

If we want to achieve this with just `-obfl-keep`, we'd probably have to wrap all the sections that we want to keep together in one single `<div class="leaf-section">`, which would change the structure of the document:

This
```xml
    <section>
        <h1>Part 1</h1>
        <section>
            <h2>Chapter 1</h2>
            <p>abcd</p>
            <section>
                <h3>Chapter 1.1</h3>
                <p>efgh</p>
            </section>
        </section>
    </section>
```

might have to become something like this:
```xml
    <section>
        <div class="leaf-section">
            <h1>Part 1</h1>
            <section>
                <h2>Chapter 1</h2>
                <p>abcd</p>
            </section>
        </div>
        <section>
            <section>
                <h3>Chapter 1.1</h3>
                <p>efgh</p>
            </section>
        </section>
    </section>
```

**Comment by bertfrees** [comment]
That won't work either because in OBFL your wrapper will be split into two sequences again (and the keep-volume property needs to be put on a block which is always a descendant of a sequence).

I talked about it today with Joel. He thinks the scope of keep-with-next-sheets should even be more narrowed, i.e. should not work beyond block boundaries. See https://github.com/brailleapps/dotify.formatter.impl/issues/40. The real issue is that we can't put the "Part 1" heading in the same block as the "Chapter 1" section, and this is because I'm using sequence breaks to jump to a new sheet. So we decided that should be tackled instead. See https://github.com/braillespecs/obfl/issues/73.
